media/js/privacy/firefox-privacy.js

```diff
@@ -0,0 +1,288 @@
+     * "Firefox by default shares data to".
+     * @param {Object} section - The section to which the widget
+     * will be added
+     */
+    function addDataChoicesWidget(section) {
+        var container = document.createElement('div');
+        var copyContainer = document.createElement('p');
+        var button = document.createElement('button');
+
+        container.setAttribute('class', 'data-choices');
+
+        copyContainer.textContent = strings.dataset.choicesCopy;
+
+        button.textContent = strings.dataset.choicesButton;
+        button.setAttribute('id', 'choose');
+        button.setAttribute('type', 'button');
+
+        container.appendChild(copyContainer);
+        container.appendChild(button);
+
+        section.appendChild(container);
+    }
+
+    /**
+     * For each main section, this innjects a button to either,
+     * show all topics under the section, or collapse all.
+     */
+    function addMasterToggles() {
+        var toggle = document.createElement('button');
+
+        toggle.classList.add('toggle');
+        toggle.setAttribute('type', 'button');
+        toggle.textContent = tabpanelOpenText;
+
+        dataAttributeHandler({
+            elem: toggle,
+            property: 'data-is-expanded',
+            value: false
+        });
+
+        for (var i = 0, l = topicHeaders.length; i < l; i++) {
+            var currentHeading = topicHeaders[i].querySelector('header h2');
+
+            dataAttributeHandler({
+                elem: toggle,
+                property: 'data-parent-container',
+                value: topicHeaders[i].id
+            });
+
+            currentHeading.insertAdjacentHTML('beforeend', toggle.outerHTML);
+        }
+    }
+
+    /**
+     * Collects all of the individual topics(sections) into an Array
+     * for use in various other functions.
+     */
+    function collectAllTopics() {
+        for (var i = 0, l = topicHeaders.length; i < l; i++) {
+            var nestedTopics = topicHeaders[i].querySelectorAll('section');
+            /* nestedTopics is a NodeList which is Array like but,
+            not a real array. We therefore need to coherce it into one. */
+            topics.push([].slice.call(nestedTopics));
+        }
+
+        /* At this point `topics` is a two dimensional array.
+        The final step then is to flatten it. */
+        topics = topics.reduce(function(a, b) {
+            return a.concat(b);
+        });
+    }
+
+    /**
+     * Simple pollyfil for setting and getting data attributes
+     * across browsers.
+     * @param {Object} data - An object containing the data needed
+     * to get or set the data attribute. For example:
+     *
+     * {
+     *     elem: document.getElementById('strings'),
+     *     property: 'isExpanded',
+     *     value: 'value'
+     * }
+     *
+     * @returns For get operations, it returns the value of the data attribute
+     */
+    function dataAttributeHandler(data) {
```

Comment:
This seems to work fine and I'm all for using newer features, but could all this somewhat complex code be avoided by just using `getAttribute` and `setAttribute`? I know they are older, but less code and greater simpliciy has some benefit still. You by no means have to change this now that it's written though :)

Reply:
Yeah, I considered that but then, the temptation of writing something to make it work won ;p